builder: fix cache paths in `-size=full` output

aykevl opened the pull request:
This fixes long paths from the TinyGo cached GOROOT, as can be seen here in the `-size=full` output:

```
   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
      0       5       0       5 |       5       5 | (padding)
    148       0       0       5 |     148       5 | (unknown)
     76       0       0       0 |      76       0 | /home/ayke/.cache/tinygo/goroot-ce8827882be9dc201bed279a631881177ae124ea064510684a3cf4bb66436e1a/src/device/arm
      4       0       0       0 |       4       0 | /home/ayke/.cache/tinygo/goroot-ce8827882be9dc201bed279a631881177ae124ea064510684a3cf4bb66436e1a/src/internal/task
```

They're now attributed to the correct package instead (device/arm and internal/task).

I've also added a test to make sure this doesn't happen again (at least not for a small `-target=microbit` binary).

aykevl committed:
builder: fix cache paths in -size=full output

This fixes long paths from the TinyGo cached GOROOT, as can be seen
here:

   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
      0       5       0       5 |       5       5 | (padding)
    148       0       0       5 |     148       5 | (unknown)
     76       0       0       0 |      76       0 | /home/ayke/.cache/tinygo/goroot-ce8827882be9dc201bed279a631881177ae124ea064510684a3cf4bb66436e1a/src/device/arm
      4       0       0       0 |       4       0 | /home/ayke/.cache/tinygo/goroot-ce8827882be9dc201bed279a631881177ae124ea064510684a3cf4bb66436e1a/src/internal/task

They're now attributed to the correct package instead (device/arm and
internal/task).

aykevl committed:
builder: add testing for -size=full

This helps to make sure this feature continues to work and we won't
accidentally introduce regressions.

aykevl commented:
I need to look into some Windows failures, probably something with `/` vs `\`.

dgryski commented:
Yup.
```
--- FAIL: TestSizeFull (0.00s)
    --- FAIL: TestSizeFull/microbit (4.59s)
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\llvm-project\compiler-rt\lib\builtins\arm
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\src\device\arm
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\src\internal\task
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\src\runtime
    --- FAIL: TestSizeFull/wasip1 (4.[7](https://github.com/tinygo-org/tinygo/actions/runs/11932446672/job/33257420141?pr=4617#step:18:8)8s)
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\src\internal\task
        sizes_test.go:112: unexpected package name in size output: D:\\a\\tinygo\tinygo\src\runtime
        sizes_test.go:112: unexpected package name in size output: D:\a\tinygo\tinygo\lib\wasi-libc\libc-bottom-half\sources
        sizes_test.go:112: unexpected package name in size output: D:\a\tinygo\tinygo\lib\wasi-libc\libc-top-half\musl\src\exit
        sizes_test.go:[11](https://github.com/tinygo-org/tinygo/actions/runs/11932446672/job/33257420141?pr=4617#step:18:12)2: unexpected package name in size output: D:\a\tinygo\tinygo\lib\wasi-libc\libc-top-half\musl\src\string
```

aykevl committed:
builder: work around bug in DWARF paths in Clang

See bug: https://github.com/llvm/llvm-project/issues/117317

aykevl commented:
Took me a while to figure out, but it's actually a Clang bug: https://github.com/llvm/llvm-project/issues/117317
I have pushed a workaround.

aykevl commented:
Apparently that wasn't the only issue, investigating the wasip1 issue...

aykevl committed:
builder: fix wasi-libc path names on Windows with -size=full

Without this fix, wasi-libc is listed as follows:

     65       0       0       0 |      65       0 | C:\Users\Ayke\src\tinygo\tinygo\lib\wasi-libc\libc-bottom-half\sources
     14       0       0       0 |      14       0 | C:\Users\Ayke\src\tinygo\tinygo\lib\wasi-libc\libc-top-half\musl\src\exit
   1398       0       0       0 |    1398       0 | C:\Users\Ayke\src\tinygo\tinygo\lib\wasi-libc\libc-top-half\musl\src\string
   1525       0       0       0 |    1525       0 | C:\Users\Ayke\src\tinygo\tinygo\lib\wasi-libc\libc-top-half\sources

With this fix, it's identified as the wasi-libc C library:

   3002       0       0       0 |    3002       0 | C wasi-libc

aykevl commented:
Should be fixed now.

deadprogram commented:
Thanks for the collection of fixes for this one @aykevl now merging.
